packages/rich-text-editor/src/CKEditor.js

```diff
@@ -0,0 +1,333 @@
+        p { margin: 0px; line-height: auto; }
+        img { width: 1.3em; height: 1.3em; }
+    `;
+
+    static externalCss = `
+        .cke_top {
+            padding: 0 !important;
+        }
+
+        .cke_toolgroup {
+            margin: 0;
+        }
+
+        .cke_dialog {
+            z-index: 20000 !important;
+        }
+
+        .cke_smiley .cke_dialog_footer {
+            display: none;
+        }
+
+        .cke_smiley .cke_dialog_title,
+        .cke_smiley .cke_dialog_close_button {
+            display: none;
+        }
+
+        .cke_dialog_background_cover {
+            opacity: 0 !important;
+            z-index: 19500 !important;
+        }
+
+        a.cke_button_off.active {
+            background: #e5e5e5;
+            border: 1px #bcbcbc solid;
+            padding: 3px 5px;
+        }
+
+        .cke_dialog_contents_body {
+            padding: 9px 10px 9px 10px !important;
+            height: auto !important;
+        }
+
+        .cke_smiley .cke_dialog_contents_body {
+            width: 150px !important;
+        }
+    `;
+
+    constructor(props) {
+        super(props);
+        this._onDocumentClick = this._onDocumentClick.bind(this);
+        this.setContainerRef = this.setContainerRef.bind(this);
+    }
+
+    componentWillReceiveProps(newProps) {
+        if (this.editor && newProps.refresh !== this.props.refresh) {
+            this.editor.setData(newProps.initialContent);
+            const wrappedEditor = new WrappedEditor(this.editor);
+            wrappedEditor.setCursorAtEnd();
+        }
+    }
+
+    componentDidMount() {
+        const {
+            onEditorChange = noop,
+            onEditorInitialized = noop,
+            options = {},
+            onEditorKey,
+        } = this.props;
+        const { editorCss } = this.constructor;
+
+        if (!window.CKEDITOR) {
+            log.error(
+                'CKEDITOR namespace can not be found on the window. You probably forgot to load the CKEditor script'
+            );
+        }
+
+        if (!CKEDITOR.getCss().includes(editorCss)) {
+            CKEDITOR.addCss(editorCss);
+        }
+
+        const fullOptions = { ...this.constructor.defaultOptions, ...options };
+        this.editor = window.CKEDITOR.replace(this.editorContainer, fullOptions);
+
+        this.editor.setData(this.props.initialContent);
+        this.editor.on('dialogShow', this._onDialogShow.bind(this));
+        this.editor.on('dialogHide', this._onDialogHide.bind(this));
+        this.editor.on('change', () => onEditorChange(this.editor.getData()));
+        if (onEditorKey) this.editor.on('key', this._onEditorKey.bind(this, onEditorKey));
+
+        // Callback to the parent to pass the editor instance so the parent can call functions on it like insertHTML.
+        onEditorInitialized(new WrappedEditor(this.editor));
+    }
+
+    _onEditorKey(onEditorKey, ev) {
```

Comment:
Do we follow this underscore convention elsewhere in d2-ui? If not, then we should try to stay consistent, and not use it.